src/wobbly/models.py

```diff
@@ -1,20 +1,467 @@
-"""Models for wobbly."""
+"""Models for Wobbly."""
 
-from pydantic import BaseModel, Field
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Annotated, Any, Literal, TypeAlias
+
+from pydantic import BaseModel, Field, PlainSerializer
 from safir.metadata import Metadata as SafirMetadata
+from safir.pydantic import SecondsTimedelta
+from vo_models.uws.types import ErrorType, ExecutionPhase
+
+JobParameters: TypeAlias = dict[str, Any] | list[str]
+"""Possible types of job parameters.
 
-__all__ = ["Index"]
+This can either be a serialized parameters model (the `dict` case), or a list
+of old-style input parameters, which are stored as simple strings.
+"""
+
+__all__ = [
+    "Index",
+    "Job",
+    "JobBase",
+    "JobCreate",
+    "JobError",
+    "JobIdentifier",
+    "JobParameters",
+    "JobResult",
+    "JobUpdate",
+    "JobUpdateAborted",
+    "JobUpdateCompleted",
+    "JobUpdateExecuting",
+    "JobUpdateError",
+    "JobUpdateMetadata",
+    "JobUpdateQueued",
+]
 
 
 class Index(BaseModel):
-    """Metadata returned by the external root URL of the application.
-
-    Notes
-    -----
-    As written, this is not very useful. Add additional metadata that will be
-    helpful for a user exploring the application, or replace this model with
-    some other model that makes more sense to return from the application API
-    root.
-    """
+    """Metadata returned by the external root URL of the application."""
 
     metadata: SafirMetadata = Field(..., title="Package metadata")
+
+
+@dataclass
+class JobIdentifier:
+    """Information required to identify a unique job.
+
+    This always includes service information. Owner information is optional
+    but enforced if present. In other words, if an owner is specified and the
+    job exists but doesn't match that owner, it is treated as if it doesn't
+    exist.
+    """
+
+    service: str
+    """Service that owns the job."""
+
+    id: str
+    """Identifier of the job."""
+
+    owner: str | None = None
+    """User who owns the job."""
+
+
+class JobError(BaseModel):
+    """Failure information about a job."""
+
+    type: Annotated[
+        ErrorType,
+        Field(
+            title="Error type",
+            description="Type of the error",
+            examples=[ErrorType.TRANSIENT, ErrorType.FATAL],
+        ),
+    ]
+
+    code: Annotated[
+        str,
```

Comment:
Sadly, there is not, because currently every IVOA standard defines its own codes and new IVOA standards may have new, different codes. In the current Safir code, this is an enum, but I realized that those were just the SODA error codes and don't work for other protocols, so I changed it to a `str` here. We'll need for the IVOA to adopt some sort of error standard to do better, alas.